apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+import * as Immutable from 'immutable';
+import Observer from '../../../Observer';
+import {update as updateWatchExpressions} from '../../../redux/watchedExpressions';
+import CommandHistory from './CommandHistory';
+import JSInterpreter from '../../../JSInterpreter';
+import watchedExpressions from '@cdo/apps/redux/watchedExpressions';
+
+const WATCH_TIMER_PERIOD = 250;
+const INITIALIZE = 'jsdebugger/INITIALIZE';
+const ATTACH = 'jsdebugger/ATTACH';
+const DETACH = 'jsdebugger/DETACH';
+const MIRROR_INTERPRETER_STATE = 'jsdebugger/MIRROR_INTERPRETER_STATE';
+const APPEND_LOG = 'jsdebugger/APPEND_LOG';
+const CLEAR_LOG = 'jsdebugger/CLEAR_LOG';
+
+// State model
+const JSDebuggerState = Immutable.Record({
+  jsInterpreter: null,
+  runApp: null,
+  observer: null,
+  watchIntervalId: null,
+  commandHistory: null,
+  logOutput: '',
+  jsInterpreterNextStep: null,
+  jsInterpreterPaused: false,
+});
+
+export function getRoot(state) {
+  return state.jsdebugger;
+}
+
+export function getCommandHistory(state) {
+  return getRoot(state).commandHistory;
+}
+
+export function getJSInterpreter(state) {
+  return getRoot(state).jsInterpreter;
+}
+
+function getRunApp(state) {
+  return getRoot(state).runApp;
+}
+
+export function isPaused(state) {
+  return getRoot(state).jsInterpreterPaused;
+}
+
+function getObserver(state) {
+  return getRoot(state).observer;
+}
+
+function getWatchIntervalId(state) {
+  return getRoot(state).watchIntervalId;
+}
+
+export function isAttached(state) {
+  return !!getJSInterpreter(state);
+}
+
+export function canRunNext(state) {
+  return (
+    isPaused(state) &&
+    getRoot(state).jsInterpreterNextStep === JSInterpreter.StepType.RUN
+  );
+}
+
+export function getLogOutput(state) {
+  return getRoot(state).logOutput;
+}
+
+export const selectors = {
+  getRoot,
+  getCommandHistory,
+  getJSInterpreter,
+  isPaused,
+  isAttached,
+  canRunNext,
+  getLogOutput,
+};
+
+// actions
+export function initialize({runApp}) {
+  return {type: INITIALIZE, runApp};
+}
+
+export function appendLog(output) {
+  return {type: APPEND_LOG, output};
+}
+
+export function clearLog() {
+  return {type: CLEAR_LOG};
+}
+
+function mirrorInterpreterState(jsInterpreter) {
+  return {type: MIRROR_INTERPRETER_STATE, jsInterpreter};
```

Comment:
Well, I think in theory your UI and your interpreter are totally decoupled from each other, and only connected through a well defined interface. That would enable you to swap out different interpreter implementations depending on your needs (like perhaps you have interpreters for different languages), while still using the same UI regardless of the underlying interpreter.

Whatever interpreter you happen to use keeps track of it's internal state however it likes. Meanwhile, your UI has some state of it's own, which is needed to actually render itself. The way you choose to model the state for the UI and the way you choose to model the state for the interpreter might differ significantly, and as such, the ui state and the interpreter state should be decoupled.

In that model, it actually makes a lot of sense to "mirror" the state in two places, although what you are really doing is mapping the interpreter's state model (or some portion of it) onto the UI's state model. So this design that I have here is actually making more sense to me now, and I think I just need to rename some things to make it more clear what is going on. I also need to hide the interpreter instance from the UI components, as they shouldn't know about the interpreter's implementation at all.